lib/matplotlib/dates.py

```diff
@@ -770,6 +770,150 @@ def __call__(self, x, pos=0):
         return cbook.unicode_safe(dt.strftime(self.fmt))
 
 
+class NewDateFormatter(ticker.Formatter):
+    """
+    This class attempts to figure out the best format to use.  This is
+    most useful when used with the :class:`AutoDateLocator`.
+
+    >>> locator = AutoDateLocator()
+    >>> formatter = NewDateFormatter(locator)
+
+    """
+
+    def __init__(self, locator, tz=None, defaultfmt='%Y-%h-%d %H:%M:%S'):
+        """
+        Autoformat the date labels.  The default format is used to form an
+        initial string, and then redundant elements are removed.
+        """
+        self._locator = locator
+        self._tz = tz
+        self.defaultfmt = defaultfmt
+        self._oldticks = np.array([])
+        self._oldlabels = None
+        self._formatter = DateFormatter(self.defaultfmt, tz)
+        self._yearfmt = '%Y'
+        self._monthfmt = '%h'
```

Comment:
looks like this should be %b, per `man strftime`:
```
 %b     The abbreviated month name according to the current locale.  (Calculated from tm_mon.)
 %h     Equivalent to %b.  (SU)  // Single Unix spec
```
the docs on python.org only mention %b

Reply:
Oh, cool, thanks.  

I'll need an API to pass these formats as well so that people can customize.